### PR TITLE
Creating Go Client Libraries for KFservice CRD

### LINKED ISSUE #161: Proposal: Add Go Client Libraries fo KFservice
/kind feature

Should we autogen client libraries for our types in Go?
This will allow users of KfServing to easily manipulate kfservices.

There seems to be an open issue for this on kubebuilder https://github.com/kubernetes-sigs/kubebuilder/issues/403 which suggests as the k8s Go client has dynamic-client then this is not required. 

It still seems to me to have static exposed client informers etc would still be useful and easy to add/update as per other libraries such as knative do.

### PR DESCRIPTION
**Pull request by SachinVarghese**
**What this PR does / why we need it**:

This PR is for creating a static go client for KFservice CRD. This has a script to create a go client from the KFService types file using the kubernetes [code-generator](https://github.com/kubernetes/code-generator/tree/kubernetes-1.12.6) that can be used to create custom informers and other kubernetes like CRUD API for KFservice CRD.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #161 

**Special notes for your reviewer**:

This PR changes the Gopkg.toml file to avoid pruning some non go files for the code generator package. Please run dep ensure before generating the client files using the hack shell script.

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
None
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/177)

### CODE REVIEW

**Comment by yuzisun** [comment]
Looks cool, @SachinVarghese  where do you intend to use this client go library ?

**Comment by ellis-bigelow** [comment]
> Please run dep ensure before generating the client files using the hack shell script.

Should we include this in the hack script then? This isn't a good thing for developers to remember. These scripts should be self contained

**Comment by cliveseldon** [comment]
@ellis-bigelow I think @SachinVarghese means that he has not included updates to the vendor folder needed for the PR as per our practice this needs to be done by a project admin.

**Comment by cliveseldon** [comment]
@yuzisun @SachinVarghese works at Seldon and plans to use it for our enterprise UI to ML deployment management.

**Comment by SachinVarghese** [comment]
> > Please run dep ensure before generating the client files using the hack shell script.
> 
> Should we include this in the hack script then? This isn't a good thing for developers to remember. These scripts should be self contained

@ellis-bigelow the `dep ensure` command is for the project admin to get the vendor updates for this PR and is not required to generate the go client files every time, just the hack script is enough.
Also, I see that there are failed tests in this PR but could not find info on what exactly went wrong. It would be great to get some more details on this.

**Comment by ellis-bigelow** [comment]
> > > Please run dep ensure before generating the client files using the hack shell script.
> > 
> > 
> > Should we include this in the hack script then? This isn't a good thing for developers to remember. These scripts should be self contained
> 
> @ellis-bigelow the `dep ensure` command is for the project admin to get the vendor updates for this PR and is not required to generate the go client files every time, just the hack script is enough.
> Also, I see that there are failed tests in this PR but could not find info on what exactly went wrong. It would be great to get some more details on this.

Ah -- yes :) I'll take care of this.

File: pkg/client/clientset/versioned/clientset.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
+/*
```
[the remark this answers is not in the document]
Comment (SachinVarghese) [reply pkg/client/clientset/versioned/clientset.go]: This is the client code generated with the code generator project. The hack script creates the pkg/client folder using the pre-generated KFService types in pkg/apis.

File: pkg/client/clientset/versioned/clientset.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	discovery "k8s.io/client-go/discovery"
+	rest "k8s.io/client-go/rest"
+	flowcontrol "k8s.io/client-go/util/flowcontrol"
+)
+
+type Interface interface {
+	Discovery() discovery.DiscoveryInterface
+	ServingV1alpha1() servingv1alpha1.ServingV1alpha1Interface
+	// Deprecated: please explicitly pick a version if possible.
+	Serving() servingv1alpha1.ServingV1alpha1Interface
```
[the remark this answers is not in the document]
Comment (SachinVarghese) [reply pkg/client/clientset/versioned/clientset.go]: The pkg/client folder is the set of client files created by the kubernetes/code-generator project. Maybe we can move to a newer version of the code-generator dependancy to improve this generated code.

**Comment by yuzisun** [comment]
> @yuzisun @SachinVarghese works at Seldon and plans to use it for our enterprise UI to ML deployment management.

@cliveseldon @SachinVarghese I am currently working on similar deployment management internally, I'd be interested in hearing how you guys are doing it.

File: hack/update-kfservice-client.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (ellis-bigelow) [reply hack/update-kfservice-client.sh]: +1. @yuzisun , can you shepherd this change once it's ready?

[the remark this answers is not in the document]
Comment (SachinVarghese) [reply hack/update-kfservice-client.sh]: @yuzisun Thanks for the reference. Will make the required changes.

[the remark this answers is not in the document]
Comment (SachinVarghese) [reply hack/update-kfservice-client.sh]: @yuzisun I have added the update and verify codegen scripts as recommended. Please review.

[the remark this answers is not in the document]
Comment (SachinVarghese) [reply hack/update-kfservice-client.sh]: @yuzisun  This client code has to be updated after every update to the CRD resulting in a change in the pkg/apis folder, how can we make it part of the same/existing process? 
And the project admin may have to run the hack script upon this PR merge to master branch to keep the client updated. Thoughts?

[the remark this answers is not in the document]
Comment (yuzisun) [reply hack/update-kfservice-client.sh]: I actually expect developer to run the hack script if he changes the CRD as part of the PR, we can add `update-codegen.sh` to current `make generate` Makefile recipe, we should also add `verify-codegen.sh` as part of CI process here (https://github.com/kubeflow/kfserving/blob/master/test/workflows/components/workflows.libsonnet#L260) to verify if the generated code is outdated , WDYT?

[the remark this answers is not in the document]
Comment (SachinVarghese) [reply hack/update-kfservice-client.sh]: @yuzisun Agreed. I have made added `update-codegen.sh` to the Makefile generate command and have added `verify-codegen.sh` as a part of the CI process. Are we good for a merge?

[the remark this answers is not in the document]
Comment (yuzisun) [reply hack/update-kfservice-client.sh]: Awesome! Looks there is an issue running the script in CI, once cleared I will approve.
http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-kfserving-presubmit-e2e-177-e7e2195-5328-3988?tab=workflow&nodeId=kubeflow-kfserving-presubmit-e2e-177-e7e2195-5328-3988-1184315456

[the remark this answers is not in the document]
Comment (SachinVarghese) [reply hack/update-kfservice-client.sh]: @yuzisun As I see, it is missing the vendor files required. You need to run the `dep ensure` to have the required vendor files to run the hack script. Refer this https://github.com/kubeflow/kfserving/pull/177#issuecomment-505532846

File: hack/update-codegen.sh [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+if [ -z "${GOPATH:-}" ]; then
+    export GOPATH=$(go env GOPATH)
+fi
+
+CODEGEN_PKG=${STARTUP_DIR}/../vendor/k8s.io/code-generator
+
+echo ${CODEGEN_PKG}
+
+${CODEGEN_PKG}/generate-groups.sh all "github.com/kubeflow/kfserving/pkg/client" "github.com/kubeflow/kfserving/pkg/apis" serving:v1alpha1
```
Comment (yuzisun) [inline hack/update-codegen.sh]: Nit: newline

File: hack/verify-codegen.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+ret=0
+diff -Naupr "${DIFFROOT}" "${TMP_DIFFROOT}" || ret=$?
+cp -a "${TMP_DIFFROOT}"/* "${DIFFROOT}"
+if [[ $ret -eq 0 ]]
+then
+  echo "${DIFFROOT} up to date."
+else
+  echo "${DIFFROOT} is out of date. Please run hack/update-codegen.sh"
+  exit 1
+fi
```
Comment (yuzisun) [inline hack/verify-codegen.sh]: Nit: newline

**Pull request merged**